algebra/euclidean_domain.lean

```diff
@@ -0,0 +1,316 @@
+  calc 
+    a%b = quotient a b * b + a%b - quotient a b * b : by ring
+    ... = a - quotient a b * b : by {dsimp[(%)]; rw (witness a b)},
+  dsimp [(%)], rw this,
+  exact dvd_sub dvd_a (dvd_mul_of_dvd_right dvd_b (a/b)),
+end
+
+/- strong lemmas -/
+
+lemma val_lt_one (a : α) : valuation a < valuation (1:α) → a = 0 :=
+begin
+  intro a_lt,
+  cases val_mul (1:α) a,
+  { exact h },
+  {
+    simp at h,
+    have := not_le_of_lt a_lt,
+    contradiction
+  }
+end
+
+lemma val_dvd_le (a b : α) :
+  b ∣ a → a ≠ 0 → valuation b ≤ valuation a :=
+begin
+  intros b_dvd ha,
+  induction b_dvd with x hx, rw hx,
+  cases val_mul b x,
+  {
+    rw h at hx, simp at hx,
+    contradiction,
+  },
+  {
+    exact h,
+  }
+end
+
+@[simp] lemma mod_one (a : α) : a % 1 = 0 :=
+begin
+  cases val_mod a 1,
+  { 
+    have := one_ne_zero, 
+    contradiction,
+  },
+  {
+    exact val_lt_one (a % 1) h,
+  }
+end 
+
+@[simp] lemma zero_mod (b : α) : 0 % b = 0 :=
+begin
+  have h1 := witness 0 b,
+  have h2 : remainder 0 b = b * (-quotient 0 b ), from
+    calc
+      remainder 0 b = quotient 0 b * b + remainder 0 b + b * (-quotient 0 b ) : by ring
+      ...                       = b * (-quotient 0 b ) : by rw [h1, zero_add],
+  cases val_mul b (-quotient 0 b),
+  {
+    simp at h, rw h at h1, simp at h1,
+    exact h1,
+  },
+  {
+    rw [←h2] at h,
+    cases val_mod 0 b,
+    {
+      rw h_1, exact mod_zero (0:α),
+    },
+    {
+      have := not_le_of_lt h_1,
+      contradiction,
+    }
+  }
+end
+
+@[simp] lemma zero_div (b : α) (hnb : b ≠ 0) : 0 / b = 0 :=
+begin
+  have h1 : remainder 0 b = 0, from zero_mod b,
+  have h2 := witness 0 b,
+  rw h1 at h2,
+  simp at h2,
+  cases eq_zero_or_eq_zero_of_mul_eq_zero h2,
+  { exact h },
+  { contradiction }
+end
+
+@[simp] lemma mod_self (a : α) : a % a = 0 :=
+begin
+  have := witness a a,
+  cases (dvd_mod_self a) with m a_mul,
+  cases val_mul a m,
+  {
+    rw [h, mul_zero] at a_mul,
+    exact a_mul,
+  },
+  {
+    cases  val_mod a a, 
+    {rw h_1, exact mod_zero (0:α)},
+    {
+      have := not_le_of_lt h_1,
+      dsimp [(%)] at a_mul,
+      rw a_mul at this,
+      contradiction,
+    }
+  }
+end 
+
+lemma div_self (a : α) : a ≠ 0 → a / a = (1:α) :=
+begin
+  intro hna,
+  have wit_aa := witness a a,
+  have a_mod_a := mod_self a, 
+  dsimp [(%)] at a_mod_a,
+  rw a_mod_a at wit_aa, simp at wit_aa,
+  have h1 : 1 * a = a, from one_mul a,
+  conv at wit_aa {for a [4] {rw ←h1}},
+  exact eq_of_mul_eq_mul_right hna wit_aa
+end
+
+/- weak gcd lemmas -/
+
+@[simp] theorem gcd_zero_left (a : α) : gcd 0 a = a := 
+begin
+  rw gcd,
+  simp,
+end
+
+@[elab_as_eliminator]
+theorem gcd.induction {P : α → α → Prop}
+                  (a b : α)
+                  (H0 : ∀ x, P 0 x)
+                  (H1 : ∀ a b, a ≠ 0 → P (b%a) a → P a b) :
+                P a b := 
+@well_founded.induction _ _ (has_well_founded.wf α) (λa, ∀b, P a b) a (λc IH,
+  by {cases decidable.em (c=0), rw h, exact H0,
```

Comment:
use `begin .. end` and spread out to multiple lines.
also `by_cases c = 0, ...` and put `{ ... }` around the tactics for both cases.

Reply:
I was following the structure of the definition of `nat.gcd.induction` (`lean/library/init/data/nat/gcd.lean`) here.
I agree agree spreading it out makes it much more readable though.